feat(parse): add `text::simple_extensions::ArgumentsItem` parser

shanretoo opened the pull request:
Adds a parser for `text::simple_extensions::ArgumentsItem` that parses the argument.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/substrait-io/substrait-rs?pullRequest=185) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/substrait-io/substrait-rs?pullRequest=185) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/substrait-io/substrait-rs?pullRequest=185) it.</sub>

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,512 @@
+/// Arguments that refer to a data value.
+#[derive(Clone, Debug)]
+pub struct ValueArgument {
+    /// A human-readable name for this argument to help clarify use.
+    name: String,
+
+    /// Additional description for this argument.
+    description: String,
```
These fields aren't required:
```suggestion
    /// A human-readable name for this argument to help clarify use.
    name: Option<String>,

    /// Additional description for this argument.
    description: Option<String>,
```

shanretoo replied:
For strings, I think we can use empty string if this field is absent to reduce the code complexity, for example, a user may need to check if the field is `None` or `Some("")`. What's your opinion here?

mbrobbel replied:
For these fields I think we should keep the `Option`. I think it's the better way of capturing that these fields are not specified in the definition. Maybe they want to follow the docs to use `arg0`, `arg1` etc here and in that case I think using `Option` makes it clear that they should handle the `None` case (if they care about it). We could also choose to reject `Some("")` or map that `None` to make it easier for consumers?

shanretoo replied:
Sounds good. I'll keep the `Option`, and reject `Some("")`.

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,512 @@
+/// Arguments that are used only to inform the evaluation and/or type derivation of the function.
+#[derive(Clone, Debug, PartialEq)]
+pub struct TypeArgument {
+    /// A human-readable name for this argument to help clarify use.
+    name: String,
+
+    /// Additional description for this argument.
+    description: String,
```
These fields aren't required:
```suggestion
    /// A human-readable name for this argument to help clarify use.
    name: Option<String>,

    /// Additional description for this argument.
    description: Option<String>,
```

shanretoo replied:
same as above.

shanretoo commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,512 @@
+        description: Option<String>,
+        options: Vec<String>,
+    ) -> Result<EnumArgument, ArgumentsItemError> {
+        if options.iter().any(String::is_empty) {
```
I didn't found this constraint in the doc. If you are sure about this, I will fix it.

mbrobbel replied:
I'm not sure about this. The schema isn't strict about it, but I wonder if it makes sense to have enum args that have no "option"? I check the extensions in https://github.com/substrait-io/substrait/tree/main/extensions, and all enum args have non-empty options e.g. https://github.com/substrait-io/substrait/blob/de4fcbc066315ca874ca163275affdc4156e570e/extensions/functions_datetime.yaml#L71-L72. Maybe we should propose a change to the schema to require at least one option?

shanretoo replied:
After looking through the doc again, I agree with you that `options` should not be empty, otherwise, this argument serves no purpose.
Also in `substrait-java`([link](https://github.com/substrait-io/substrait-java/blob/79decd20e85d6a1a5623890042ebcf1415cf784a/core/src/main/java/io/substrait/expression/EnumArg.java#L25)), it implicitly requires `options` is not empty.

shanretoo commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,512 @@
+    /// Whether this argument is required to be a constant for invocation.
+    /// For example, in some system a regular expression pattern would only be accepted as a literal
+    /// and not a column value reference.
+    constant: bool,
```
As specified in doc, this defaults to false if absent. I think we can remove the option here?

mbrobbel replied:
This would make the back-conversion lossy, however, this is fine because it doesn't change behaviour. And I guess it makes it easier for users to not have to be aware of this (default=false) detail.

shanretoo replied:
Make sense.

shanretoo commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,512 @@
+}
+
+impl ValueArgument {
+    fn parse<C: Context>(
```
same as above.

shanretoo commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,512 @@
+}
+
+impl TypeArgument {
+    fn parse<C: Context>(
```
same as above.

shanretoo commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,512 @@
+/// Arguments that support a fixed set of declared values as constant arguments.
+#[derive(Clone, Debug, PartialEq)]
+pub struct EnumArgument {
+    /// A human-readable name for this argument to help clarify use.
+    name: String,
+
+    /// Additional description for this argument.
+    description: String,
```
same as above.

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,512 @@
+}
+
+impl EnumArgument {
+    fn parse(
```
The approach you could use here is: implement `Parse` for the `ArgumentsItem` that forwards the parser to proceed parsing based on the variant. Those `Parse` implementations are parsers for the data structs of the variants (`EnumArgument`, `ValueArgument`, `TypeArgument`). I followed a similar approach for enums [for the `MappingType` enum (this was part of an unmerged PR)](https://github.com/substrait-io/substrait-rs/blob/aebd6cc53152cda3602decfbd096329e6f4e5850/src/parse/extensions/simple_extension_declaration/mod.rs).

shanretoo replied:
I'm still confused here.

The structure of `ArgumentsItem` contrasts with that of `MappingType`, as shown below. In `ArgumentsItem`, the enum fields are struct-like enum variants, whereas in `MappingType`, the enum fields consist of single-field tuples. So it is impossible to use the same approach that forwards the parser to proceed parsing based on the variant for `ArgumentsItem`. 

```
pub enum ArgumentsItem {
    Variant0 {
        #[serde(default, skip_serializing_if = "Option::is_none")]
        description: Option<String>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        name: Option<String>,
        options: Vec<String>,
    },
```
```
pub enum MappingType {
    #[prost(message, tag = "1")]
    ExtensionType(ExtensionType),
    #[prost(message, tag = "2")]
    ExtensionTypeVariation(ExtensionTypeVariation),
    #[prost(message, tag = "3")]
    ExtensionFunction(ExtensionFunction),
}
```

For `MappingType`, the code works:
```
match self {
    Self::ExtensionType(extension_type) => ctx.parse_record_value(extension_type)
```

However, for `ArgumentsItem`, the code throws compilation error:
```
error[E0164]: expected tuple struct or tuple variant, found struct variant `Self::Variant0`
  --> src/parse/text/simple_extensions/argument.rs:53:13
   |         match self {
53 |             Self::Variant0(enum_argument) => ctx.parse(enum_argument),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a tuple struct or tuple variant
```

So that, I have to destruct all the fields from the struct enum field and pass them to a `parse` function:
```
match self {
    simple_extensions::ArgumentsItem::Variant0 {
        name,
        description,
        options,
    } => EnumArgument::parse(name, description, options)
```

Correct me if I'm wrong.

mbrobbel replied:
Oh right 🤦. Let me think a bit about this. Maybe we should try to change how these enums are generated by `typify`.

shanretoo replied:
It would be the best if we can change these enums.

mbrobbel replied:
With the changes in https://github.com/substrait-io/substrait/pull/642, `typify` generates:
```rust
pub enum ArgumentsItem {
    EnumerationArg(EnumerationArg),
    ValueArg(ValueArg),
    TypeArg(TypeArg),
}

pub struct EnumerationArg {
    pub description: Option<String>,
    pub name: Option<String>,
    pub options: Vec<String>
}
```

shanretoo replied:
I will fix them once it merged.

mbrobbel replied:
You can merge/rebase `main`.

mbrobbel commented:
> How about adding tests against the extensions in the substrait submodule in the future prs, i.e. after including extensions in this repo like the work you have done in the closed pr: https://github.com/mbrobbel/substrait-rs/blob/aebd6cc53152cda3602decfbd096329e6f4e5850/build.rs#L188-L197 @mbrobbel

Sounds good: https://github.com/substrait-io/substrait-rs/issues/186

shanretoo commented:
@mbrobbel 

I have implemented `Parse` for those enums and fixed corresponding tests.

But when I try to add checks to ensure the parsing work for core extensions, I encountered a problem:
1. The `parse` method of `Parse` trait is using the value receiver of `self`.
```rust
fn parse(self, ctx: &mut C) -> Result<Self::Parsed, Self::Error>;
```
2. But the `ArgumentsItem`s extracted from `EXTENSIONS` are references and `ArgumentsItem` does not implement `Copy` trait. It is not possible to call `parse` for them.
```rust
error[E0507]: cannot move out of `*item` which is behind a shared reference
   --> src/parse/text/simple_extensions/argument.rs:632:21
    |
632 |                     item.parse(&mut Context::default())?;
    |                     ^^^^ move occurs because `*item` has type `text::simple_extensions::ArgumentsItem`, which does not implement the `Copy` trait
```
I'm wondering that if we can modify `parse` method to use the reference receiver.
What's your opinion here?

mbrobbel commented:
>     2. But the `ArgumentsItem`s extracted from `EXTENSIONS` are references and `ArgumentsItem` does not implement `Copy` trait. It is not possible to call `parse` for them.
> 

You can [clone](https://docs.rs/substrait/latest/substrait/text/simple_extensions/struct.SimpleExtensions.html#impl-Clone-for-SimpleExtensions) the `SimpleExtensions` before parsing them.

> I'm wondering that if we can modify `parse` method to use the reference receiver. What's your opinion here?

We could do this, but it moves cloning (where needed) into the parser impls - unless we also change all parsed items to reference data in the original items. I think the common case would be for users to have owned instances they want to parse, and when they don't, like in your example when you only have a reference, users can explicitly clone. What do you think?

shanretoo commented:
Thank you for your detailed explanation. It makes sense to me. This pr is ready for review.

mbrobbel commented on src/parse/text/simple_extensions/mod.rs:
```diff
@@ -9,6 +9,9 @@ use crate::{
     text,
 };
 
+#[allow(missing_docs)]
```
Nit: can you add a simple doc string to the module?

shanretoo replied:
Added in [commit](https://github.com/substrait-io/substrait-rs/pull/185/commits/42738efd2bab1e5f961594ddfe30d02d5ff0c52d).

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,633 @@
+}
+
+impl ArgumentsItem {
+    #[inline]
+    fn parse_name(name: Option<String>) -> Result<Option<String>, ArgumentsItemError> {
+        match name {
+            Some(s) if s.is_empty() => {
+                Err(ArgumentsItemError::EmptyOptionalField("name".to_string()))
+            }
+            _ => Ok(name),
+        }
+    }
+
+    #[inline]
+    fn parse_description(
+        description: Option<String>,
+    ) -> Result<Option<String>, ArgumentsItemError> {
+        match description {
+            Some(s) if s.is_empty() => Err(ArgumentsItemError::EmptyOptionalField(
+                "description".to_string(),
+            )),
+            _ => Ok(description),
+        }
+    }
```
Can we merge this and pass the name for the error as argument?

shanretoo replied:
Extracted the shared logic into a new function: [commit](https://github.com/substrait-io/substrait-rs/pull/185/commits/80873cf325f1bb026355f627003421399c9b48dc).

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,633 @@
+
+/// Arguments that support a fixed set of declared values as constant arguments.
+#[derive(Clone, Debug, PartialEq)]
+pub struct EnumerationArg {
```
Can you add an impl with methods to return references to the fields of this struct? Similar to for example: https://docs.rs/substrait/latest/substrait/parse/proto/struct.Version.html#impl-Version.

shanretoo replied:
Added in [commit](https://github.com/substrait-io/substrait-rs/pull/185/commits/4713c54f3bf70553ac9fdbe389264616fcf910e6).

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,633 @@
+    }
+}
+
+impl From<EnumerationArg> for simple_extensions::ArgumentsItem {
+    fn from(value: EnumerationArg) -> Self {
+        simple_extensions::ArgumentsItem::EnumerationArg(value.into())
+    }
+}
+
+impl From<EnumerationArg> for ArgumentsItem {
+    fn from(value: EnumerationArg) -> Self {
+        ArgumentsItem::EnumArgument(value)
+    }
+}
```
Out of curiosity: are these impls used/required?

shanretoo replied:
These are not required but we can make the type conversion simpler with them.

The former is used in:
```rust
impl From<ArgumentsItem> for simple_extensions::ArgumentsItem {
    fn from(value: ArgumentsItem) -> Self {
        match value {
            ArgumentsItem::EnumArgument(arg) => arg.into(),
            ArgumentsItem::ValueArgument(arg) => arg.into(),
            ArgumentsItem::TypeArgument(arg) => arg.into(),
        }
    }
}
```
While the latter is used in:
```rust
impl<C: Context> Parse<C> for simple_extensions::ArgumentsItem {
    type Parsed = ArgumentsItem;
    type Error = ArgumentsItemError;

    fn parse(self, ctx: &mut C) -> Result<Self::Parsed, Self::Error> {
        match self {
            simple_extensions::ArgumentsItem::EnumerationArg(arg) => Ok(ctx.parse(arg)?.into()),
            simple_extensions::ArgumentsItem::ValueArg(arg) => Ok(ctx.parse(arg)?.into()),
            simple_extensions::ArgumentsItem::TypeArg(arg) => Ok(ctx.parse(arg)?.into()),
        }
    }
}
```

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,633 @@
+
+/// List of valid string options
+#[derive(Clone, Debug, PartialEq)]
+pub struct EnumOptions(Vec<String>);
```
Should we switch this to store the `HashSet` that we build in the `Parse` impl? We lose the order, but that doesn't change the meaning. It captures that these options are unique.

shanretoo replied:
Sounds good. Fixed in [commit](https://github.com/substrait-io/substrait-rs/pull/185/commits/c16c017dce3ade6928194f65b6b30c01ce252bb2).

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,633 @@
+
+/// List of valid string options
+#[derive(Clone, Debug, PartialEq)]
+pub struct EnumOptions(Vec<String>);
```
Can you also add an impl with a method that returns the options?

shanretoo replied:
How about implementing `Deref` for it as in [commit](https://github.com/substrait-io/substrait-rs/pull/185/commits/7b639c9e16f1d2ca02c506acca83199a64f73f70)?

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,633 @@
+
+/// Arguments that refer to a data value.
+#[derive(Clone, Debug)]
+pub struct ValueArg {
```
Can you also add methods for this one?

shanretoo replied:
Added in [commit](https://github.com/substrait-io/substrait-rs/pull/185/commits/29f56d9136b6ff9db1044adb532c9e91a35071e0). The method for `value` is not added for now since it has not been parsed yet.

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,633 @@
+    /// Whether this argument is required to be a constant for invocation.
+    /// For example, in some system a regular expression pattern would only be accepted as a literal
+    /// and not a column value reference.
+    constant: Option<bool>,
```
For this one, do we want the method to return `false` if the option is `None`? Or do  we want to store a `bool` that defaults to `false` if unspecified (what you had before). This is different than the name/description fields in that the spec specifies a default when this field is unset.

shanretoo replied:
I think we can return `false` if it is `None`, so that the user does not need to handle the default value separately.

shanretoo replied:
I put the description of the default behavior in the comment: [commit](https://github.com/substrait-io/substrait-rs/pull/185/commits/9ed057879faf83f32006224c8314947a9bf4bfed).

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,633 @@
+
+/// Arguments that are used only to inform the evaluation and/or type derivation of the function.
+#[derive(Clone, Debug, PartialEq)]
+pub struct TypeArg {
```
And can you also add methods for this one?

shanretoo replied:
Added in [commit](https://github.com/substrait-io/substrait-rs/pull/185/commits/53a2a22eb4e76716ac5b016ae46274a3e5524479).

mbrobbel reviewed (commented):
Thanks! 
I have some final questions/suggestions - but we can also move some of these to follow-up PRs if you want.

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,712 @@
+impl ArgumentsItem {
+    /// Parses an `Option<String>` field, rejecting it if an empty string is provided.
+    #[inline]
+    fn parse_optinal_string(
```
```suggestion
    fn parse_optional_string(
```

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,712 @@
+
+    #[inline]
+    fn parse_name(name: Option<String>) -> Result<Option<String>, ArgumentsItemError> {
+        ArgumentsItem::parse_optinal_string("name", name)
```
```suggestion
        ArgumentsItem::parse_optional_string("name", name)
```

mbrobbel commented on src/parse/text/simple_extensions/argument.rs:
```diff
@@ -0,0 +1,712 @@
+    fn parse_description(
+        description: Option<String>,
+    ) -> Result<Option<String>, ArgumentsItemError> {
+        ArgumentsItem::parse_optinal_string("description", description)
```
```suggestion
        ArgumentsItem::parse_optional_string("description", description)
```

mbrobbel reviewed (commented):
Noticed one typo, but other than that this looks good, thanks again!

shanretoo commented:
Really appreciate your review!
